### PR TITLE
Optimization for StringBuilder append boolean and null

### PR DESCRIPTION
**Pull request by wenshao**
Improve the performance of StringBuilder's append(boolean) and append(null).

When the coder is latin1, treat null/fals/true as an int. If the coder is utf16, treat it as a long, use ByteArrayLittleEndian.setInt/setLong to write it into buf.

The performance numbers under MacBookPro M1 Max are as follows:
```diff
-Benchmark                             Mode  Cnt   Score   Error  Units
-StringBuilders.toStringCharWithBool8  avgt   15  15.766 ? 0.355  ns/op
-StringBuilders.toStringCharWithNull8  avgt   15  10.723 ? 0.019  ns/op

+Benchmark                             Mode  Cnt   Score   Error  Units
+StringBuilders.toStringCharWithBool8  avgt   15  13.023 ? 0.141  ns/op (+21.07)
+StringBuilders.toStringCharWithNull8  avgt   15  8.480  ? 0.027  ns/op (+26.46)
```

Please review and don't hesitate to critique my approach and patch.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15990/head:pull/15990` \
`$ git checkout pull/15990`

Update a local copy of the PR: \
`$ git checkout pull/15990` \
`$ git pull https://git.openjdk.org/jdk.git pull/15990/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15990`

View PR using the GUI difftool: \
`$ git pr show -t 15990`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15990.diff">https://git.openjdk.org/jdk/pull/15990.diff</a>

</details>

### CODE REVIEW

**Comment by altrisi** [comment]
Seems like you included changes from another PR here (new `AbstractStringBuilder` `append` method).

File: src/java.base/share/classes/java/lang/AbstractStringBuilder.java [hunk 1]
```diff
@@ -762,25 +802,23 @@ public AbstractStringBuilder append(boolean b) {
...
-            if (b) {
-                count = StringUTF16.putCharsAt(val, count, 't', 'r', 'u', 'e');
+            if (latin1) {
+                ByteArrayLittleEndian.setInt(val, count, Constants.FALS_LATIN1);
+                val[count + 4] = 'e';
             } else {
-                count = StringUTF16.putCharsAt(val, count, 'f', 'a', 'l', 's', 'e');
+                ByteArrayLittleEndian.setLong(val, count << 1, Constants.FALS_UTF16);
+                StringUTF16.putChar(val, count + 4, 'e');
             }
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/AbstractStringBuilder.java:820]: Are the `StringUTF16​::putCharsAt(…)` methods still used anywhere after this?

Comment (wenshao) [reply src/java.base/share/classes/java/lang/AbstractStringBuilder.java:820]: > Are the `StringUTF16​::putCharsAt(…)` methods still used anywhere after this?

It is used in the test code and should be deleted.

**Comment by wenshao** [comment]
@cl4es could you please review my code? Thank you!

**Comment by cl4es** [comment]
Are these common enough that 2-3ns/op is a real win on any real app? 

Instead of obfuscating core library code with `ByteArrayLittleEndian` I really want us to spend some time analyzing the trivial `byte[]` write operations and investigate if/how JITs can be improved here.

File: src/java.base/share/classes/java/lang/AbstractStringBuilder.java [hunk 2]
```diff
@@ -799,28 +795,30 @@ public AbstractStringBuilder append(char[] str, int offset, int len) {
      * @return  a reference to this object.
      */
     public AbstractStringBuilder append(boolean b) {
-        ensureCapacityInternal(count + (b ? 4 : 5));
         int count = this.count;
+        int spaceNeeded = count + (b ? 4 : 5);
+        ensureCapacityInternal(spaceNeeded);
         byte[] val = this.value;
         boolean latin1 = isLatin1();
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/AbstractStringBuilder.java:802]: Now unused:
```suggestion
```

**Comment by wenshao** [comment]
> Are these common enough that 2-3ns/op is a real win on any real app?
> 
> Instead of obfuscating core library code with `ByteArrayLittleEndian` I really want us to spend some time analyzing the trivial `byte[]` write operations and investigate if/how JITs can be improved here.

I think StringBuilder.append(boolean/null) are common operations, and now we know that using ByteArrayLittleEndian can improve performance. Does JIT have any plans to do such optimization?

**Comment by liach** [comment]
Can you try replacing the individual char puts with `putStringAt` with constant arguments `"null"` `"false"` `"true"` and check the performance results? I expect the benchmarks to show copying a constant byte array in full (the latin-1 bytes are stored in those string constants, not sure about how it's carried out exactly) versus putting the bytes into string with unaligned writes.

**Comment by wenshao** [comment]
> Can you try replacing the individual char puts with `putStringAt` with constant arguments `"null"` `"false"` `"true"` and check the performance results? I expect the benchmarks to show copying a constant byte array in full (the latin-1 bytes are stored in those string constants, not sure about how it's carried out exactly) versus putting the bytes into string with unaligned writes.

The implementation using append(b ? "true" : "false") will be slower than the baseline version

```java
    public AbstractStringBuilder append(boolean b) {
        if (isLatin1()) {
            return append(b ? "true" : "false");
        }
        ...
    }
```

**Comment by cl4es** [comment]
We now have a PR for merging stores efficiently: #16245 It would be good to run the microbenchmarks you have evaluated here on that PR branch and see if that optimization narrows the gap. I have some hope we'll be able to dial back the use of `ByteArray(LittleEndian)` without sacrificing performance.

**Comment by wenshao** [comment]
PR #16245 performs better

* MacBookPro M1 Pro
```
Benchmark                             Mode  Cnt   Score   Error  Units # master
StringBuilders.toStringCharWithBool8  avgt   15  16.127 ? 0.223  ns/op

Benchmark                             Mode  Cnt   Score   Error  Units # pr/15990
StringBuilders.toStringCharWithBool8  avgt   15  13.955 ? 0.181  ns/op

Benchmark                             Mode  Cnt   Score   Error  Units # pr/16245
StringBuilders.toStringCharWithBool8  avgt   15  12.519 ? 0.228  ns/op
```

File: src/java.base/share/classes/java/lang/AbstractStringBuilder.java [hunk 3]
```diff
@@ -61,6 +61,7 @@
  */
 abstract sealed class AbstractStringBuilder implements Appendable, CharSequence
     permits StringBuilder, StringBuffer {
+
```
Comment (liach) [inline src/java.base/share/classes/java/lang/AbstractStringBuilder.java:64]:
```suggestion
```
Redundant line.

**Review by liach — APPROVED** [review]

**Comment by wenshao** [comment]
The current implementation will have performance degradation when the coder is UTF16, but if combined with PR #16245, it will have very good performance.

The performance numbers under MacBookPro M1 Max are as follows:

```
Benchmark                                  Mode  Cnt    Score    Error  Units  # master
StringBuilders.toStringCharUTF16WithBool8  avgt   15   40.139 ?  0.113  ns/op
StringBuilders.toStringCharUTF16WithNull8  avgt   15   90.630 ?  5.380  ns/op
StringBuilders.toStringCharWithBool8       avgt   15   89.899 ?  2.695  ns/op
StringBuilders.toStringCharWithNull8       avgt   15   25.028 ?  0.150  ns/op

Benchmark                                  Mode  Cnt    Score    Error  Units  # PR 15990 (0d4d519) current version
StringBuilders.toStringCharUTF16WithBool8  avgt   15   44.336 ?  0.190  ns/op
StringBuilders.toStringCharUTF16WithNull8  avgt   15  126.908 ? 22.450  ns/op
StringBuilders.toStringCharWithBool8       avgt   15   64.320 ?  1.573  ns/op
StringBuilders.toStringCharWithNull8       avgt   15   23.825 ?  0.152  ns/op

Benchmark                                  Mode  Cnt    Score    Error  Units  # PR 16245
StringBuilders.toStringCharUTF16WithBool8  avgt   15   39.976 ?  0.179  ns/op
StringBuilders.toStringCharUTF16WithNull8  avgt   15   90.817 ?  5.681  ns/op
StringBuilders.toStringCharWithBool8       avgt   15   36.279 ?  1.273  ns/op
StringBuilders.toStringCharWithNull8       avgt   15   21.696 ?  0.109  ns/op

Benchmark                                  Mode  Cnt    Score    Error  Units  # PR 16245 + 15990
StringBuilders.toStringCharUTF16WithBool8  avgt   15   18.389 ?  0.408  ns/op
StringBuilders.toStringCharUTF16WithNull8  avgt   15   12.929 ?  0.023  ns/op
StringBuilders.toStringCharWithBool8       avgt   15   12.395 ?  0.176  ns/op
StringBuilders.toStringCharWithNull8       avgt   15    7.999 ?  0.109  ns/op
```

File: src/java.base/share/classes/java/lang/AbstractStringBuilder.java (hunk 3 above)
Comment (wenshao) [reply src/java.base/share/classes/java/lang/AbstractStringBuilder.java:64]: This extra line existed before this PR

**Pull request closed without merging**